File: ios/EscapePhoneTests/ConvenienceStoreThemeTests.swift
import XCTest
@testable import EscapePhone

@MainActor
final class ConvenienceStoreThemeTests: XCTestCase {
    func testSelectTheme_withTheLastCommit_opensCorrectMenu() {
        let store = InMemoryThemeProgressStore()
        var progress = store.loadThemeProgress(themeId: .theLastCommit)
        progress.themeStatus = .inProgress
        store.saveThemeProgress(progress)
        XCTAssertEqual(store.loadThemeProgress(themeId: .theLastCommit).themeId, .theLastCommit)
    }

    func testSelectTheme_withConvenienceStoreLoop_opensCorrectMenu() {
        let store = InMemoryThemeProgressStore()
        let container = ConvenienceStoreContainer(themeProgressStore: store)
        container.startNewGame()
        XCTAssertTrue(container.hasSavedGame)
        XCTAssertEqual(store.loadThemeProgress(themeId: .convenienceStoreLoop).themeId, .convenienceStoreLoop)
    }

    func testSaveAndLoad_convenienceStoreProgress_restoresState() {
        let store = InMemoryThemeProgressStore()
        let container = ConvenienceStoreContainer(themeProgressStore: store)
        container.startNewGame()
        container.completeIntro()
        container.completeReceiptPuzzle()
        let reloaded = ConvenienceStoreContainer(themeProgressStore: store)
        XCTAssertTrue(reloaded.progress.receiptSolved)
    }

    func testResetConvenienceStoreProgress_keepsTheLastCommitProgress() {
        let store = InMemoryThemeProgressStore()
        var lastCommit = store.loadThemeProgress(themeId: .theLastCommit); lastCommit.hintCount = 4
        store.saveThemeProgress(lastCommit)
        let container = ConvenienceStoreContainer(themeProgressStore: store)
        container.startNewGame()
        container.reset()
        XCTAssertEqual(store.loadThemeProgress(themeId: .theLastCommit).hintCount, 4)
    }

    func testSelectPublicDisclosure_savesConvenienceStoreEnding() {
        let store = InMemoryThemeProgressStore()
        let container = ConvenienceStoreContainer(themeProgressStore: store)
        container.startNewGame()
        container.selectPublicDisclosure()
        XCTAssertEqual(container.progress.endingType, .publicDisclosure)
    }

    func testSelectEncryptedArchive_savesConvenienceStoreEnding() {
        let store = InMemoryThemeProgressStore()
        let container = ConvenienceStoreContainer(themeProgressStore: store)
        container.startNewGame()
        container.selectEncryptedArchive()
        XCTAssertEqual(container.progress.endingType, .encryptedArchive)
    }

    func testCompleteConvenienceStoreTheme_marksThemeCompleted() {
        let store = InMemoryThemeProgressStore()
        let container = ConvenienceStoreContainer(themeProgressStore: store)
        container.startNewGame()
        container.selectPublicDisclosure()
        let theme = store.loadThemeProgress(themeId: .convenienceStoreLoop)
        XCTAssertEqual(theme.themeStatus, .completed)
        if case .convenienceStoreLoop = theme.themeSpecificState {} else { XCTFail("expected convenienceStoreLoop state") }
    }
}

File: AOS/app/src/test/java/com/example/escapephone/ConvenienceStoreThemeTests.kt
package com.example.escapephone

import com.example.escapephone.app.ConvenienceStoreViewModel
import com.example.escapephone.core.model.ConvenienceStoreEndingType
import com.example.escapephone.core.theme.InMemoryThemeProgressStore
import com.example.escapephone.core.theme.ThemeId
import com.example.escapephone.core.theme.ThemeSpecificState
import com.example.escapephone.core.theme.ThemeStatus
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class ConvenienceStoreThemeTests {
    @Test
    fun selectTheme_withTheLastCommit_opensCorrectMenu() {
        val store = InMemoryThemeProgressStore()
        store.saveThemeProgress(store.loadThemeProgress(ThemeId.theLastCommit).copy(themeStatus = ThemeStatus.inProgress))
        assertEquals(ThemeId.theLastCommit, store.loadThemeProgress(ThemeId.theLastCommit).themeId)
    }

    @Test
    fun selectTheme_withConvenienceStoreLoop_opensCorrectMenu() {
        val store = InMemoryThemeProgressStore()
        val viewModel = ConvenienceStoreViewModel(store)
        viewModel.startNewGame()
        assertTrue(viewModel.hasSavedGame())
        assertEquals(ThemeId.convenienceStoreLoop, store.loadThemeProgress(ThemeId.convenienceStoreLoop).themeId)
    }

    @Test
    fun saveAndLoad_convenienceStoreProgress_restoresState() {
        val store = InMemoryThemeProgressStore()
        val viewModel = ConvenienceStoreViewModel(store)
        viewModel.startNewGame()
        viewModel.completeIntro()
        viewModel.completeReceiptPuzzle()
        val reloaded = ConvenienceStoreViewModel(store)
        assertTrue(reloaded.uiState.value.progress.receiptSolved)
    }

    @Test
    fun resetConvenienceStoreProgress_keepsTheLastCommitProgress() {
        val store = InMemoryThemeProgressStore()
        store.saveThemeProgress(store.loadThemeProgress(ThemeId.theLastCommit).copy(hintCount = 4))
        val viewModel = ConvenienceStoreViewModel(store)
        viewModel.startNewGame()
        viewModel.reset()
        assertEquals(4, store.loadThemeProgress(ThemeId.theLastCommit).hintCount)
    }

    @Test
    fun selectPublicDisclosure_savesConvenienceStoreEnding() {
        val store = InMemoryThemeProgressStore()
        val viewModel = ConvenienceStoreViewModel(store)
        viewModel.startNewGame()
        viewModel.selectPublicDisclosure()
        val progress = viewModel.uiState.value.progress
        assertEquals(ConvenienceStoreEndingType.publicDisclosure, progress.endingType)
    }

    @Test
    fun selectEncryptedArchive_savesConvenienceStoreEnding() {
        val store = InMemoryThemeProgressStore()
        val viewModel = ConvenienceStoreViewModel(store)
        viewModel.startNewGame()
        viewModel.selectEncryptedArchive()
        val progress = viewModel.uiState.value.progress
        assertEquals(ConvenienceStoreEndingType.encryptedArchive, progress.endingType)
    }

    @Test
    fun completeConvenienceStoreTheme_marksThemeCompleted() {
        val store = InMemoryThemeProgressStore()
        val viewModel = ConvenienceStoreViewModel(store)
        viewModel.startNewGame()
        viewModel.selectPublicDisclosure()
        val theme = store.loadThemeProgress(ThemeId.convenienceStoreLoop)
        assertEquals(ThemeStatus.completed, theme.themeStatus)
        val state = theme.themeSpecificState
        assertTrue(state is ThemeSpecificState.ConvenienceStoreLoopState)
    }
}
